### DISCUSSION TITLE
How to Trigger a Job with Sensors Based on dbt Asset Job Outcomes

### DISCUSSION THREAD

**Discussion by dennisgera** [dagster-io/dagster#19874] [Q&A] [answered]
I'm seeking best practices for monitoring pipelines with dbt asset jobs, aiming to trigger a job that notifies an observability platform via webhook upon any dbt job execution, success or failure. This is vital for real-time insights into dbt test outcomes and their data ecosystem impact.

Here's the context of what I've implemented so far:

I have multiple dbt asset jobs created with create_dbt_asset_job, tagged and selected based on their relevance, such as "antifraud" and "common" assets. For example:

```python
def create_dbt_asset_job(name, dbt_assets, dbt_select, dbt_exclude=None):
    return define_asset_job(
        name=name,
        selection=build_dbt_asset_selection(dbt_assets=dbt_assets, dbt_select=dbt_select, dbt_exclude=dbt_exclude),
        tags=DBT_K8_CONFIG_TAGS,
    )

antifraud_assets_job = create_dbt_asset_job(
    name="materialize_antifraud_assets",
    dbt_assets=[arado_dbt_assets],
    dbt_select="tag:antifraud"
)

common_assets_job = create_dbt_asset_job(
    name="common_assets_job",
    dbt_assets=[arado_dbt_assets],
    dbt_select="+tag:common",
    dbt_exclude="tag:static tag:staging,tag:brain config.materialized:snapshot",
)
```
To monitor these jobs, I've set up sensors using @run_status_sensor to trigger my webhook job whenever any of these dbt jobs are completed, regardless of their outcome:

```python

def make_job_sensor(
    name: str,
    trigger_jobs,
    request_job,
    run_status: DagsterRunStatus,
    run_config=None,
    run_config_fn=None,
    default_status_sensor=None,
) -> SensorDefinition:
    if default_status_sensor is None:
        default_status_sensor = default_status.sensor

    @run_status_sensor(
        name=name,
        request_job=request_job,
        run_status=run_status,
        monitored_jobs=trigger_jobs,
        default_status=default_status_sensor,
    )
    def _job_sensor(context: RunStatusSensorContext):
        _run_config = run_config_fn(context) if run_config_fn else run_config
        yield RunRequest(run_key=f"triggered_by_{context.dagster_run.run_id}", run_config=_run_config)

    return _job_sensor


def make_successful_job_sensor(
    name, trigger_jobs, request_job, run_config=None, run_config_fn=None, default_status_sensor=None
) -> SensorDefinition:
    return make_job_sensor(
        name, trigger_jobs, request_job, DagsterRunStatus.SUCCESS, run_config, run_config_fn, default_status_sensor
    )


def make_failed_job_sensor(
    name, trigger_jobs, request_job, run_config=None, run_config_fn=None, default_status_sensor=None
) -> SensorDefinition:
    return make_job_sensor(
        name, trigger_jobs, request_job, DagsterRunStatus.FAILURE, run_config, run_config_fn, default_status_sensor
    )


dbt_jobs = [antifraud_assets_job, common_assets_job, ...]

sync_elementary_on_success = make_successful_job_sensor(
    name="sync_elementary_on_success",
    request_job=sync_elementary_data,
    trigger_jobs=dbt_jobs,
)

sync_elementary_on_failure = make_failed_job_sensor(
    name="sync_elementary_on_failure",
    request_job=sync_elementary_data,
    trigger_jobs=dbt_jobs,
)
```

While this approach works, I'm curious if there's a more streamlined or efficient way to achieve this, especially considering I have to create two separate sensors for success and failure cases for each dbt job.

My key questions are:

1. Is there a way to combine the success and failure sensors into a single sensor that triggers the job, or is my current setup the recommended practice?
2. Are there any patterns within Dagster for monitoring multiple dbt asset jobs and triggering actions based on their run status?

**Comment by yuhan** [comment, marked as answer]
> 1. Is there a way to combine the success and failure sensors into a single sensor that triggers the job, or is my current setup the recommended practice?

The benefit of separating them is that on the individual sensor page, you'd be able to clearly tell which runs fails vs succeeds from the tick history. So I don't see it being a problematic approach. However, the benefit of combining success and failure sensors is better maintainability of the code itself. While we don't have a built-in API for it, you could potentially write a custom sensor that checks the run status within the sensor function - effectively replicating the run status sensor logic in a regular sensor.

> 2. Are there any patterns within Dagster for monitoring multiple dbt asset jobs and triggering actions based on their run status?

You can use the `monitored_jobs` param in `@run_status_sensor` to monitor multiple jobs.
